lib/ember-app.js

```diff
@@ -54,30 +55,49 @@ function EmberApp(options) {
   this._app = AppFactory['default']();
 }
 
-EmberApp.prototype.boot = function() {
-  return this._app.boot();
+EmberApp.prototype.boot = function(options) {
+  return this._app.boot().then(function(app) {
+    return app.buildInstance().boot(options);
+  });
 };
 
-EmberApp.prototype.visit = function(url) {
+EmberApp.prototype.visit = function(request, response) {
+  var path = request.path;
   var doc = new SimpleDOM.Document();
   var rootElement = doc.body;
   var options = { isBrowser: false, document: doc, rootElement: rootElement };
 
-  return this._app.visit(url, options).then(function(instance) {
-    var head;
-    if (doc.head) {
-      head = HTMLSerializer.serialize(doc.head);
-    }
-    try {
-      return {
-        url: instance.getURL(), // TODO: use this to determine whether to 200 or redirect
-        title: doc.title,
-        head: head,
-        body: HTMLSerializer.serialize(rootElement) // This matches the current code; but we probably want `serializeChildren` here
-      };
-    } finally {
-      instance.destroy();
+  return this.boot(options).then(function(instance) {
+    var fastbootService = instance.lookup('service:fastboot');
+
+    var cookies = request.get('cookie');
+    if (cookies) {
+      cookies = cookie.parse(cookies);
     }
+
+    fastbootService.setProperties({
```

Comment:
@marcoow Yes, this is definitely a tradeoff of this implementation. I went spelunking in the Ember `Application` and `ApplicationInstance` code and it seems there is not an elegant way of setting cookies prior to running instance initializers.

That said, I just re-read the code now and I think I misread last night. I thought that registry creation and running of instance initializers all happened in [_bootSync](https://github.com/emberjs/ember.js/blob/master/packages/ember-application/lib/system/application-instance.js#L140-L169). But actually it looks like `__registry__` exists on the application instance before the boot process begins.

I _think_ that means that we could register an object in the registry before the instance initializers are kicked off. It involves reaching pretty deep into private API, so we should think about what we're trying to express here and see if there's an elegant way of exposing it as public API.